channels/rail/server/rail_main.c

```diff
@@ -0,0 +1,1616 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * RAIL Virtual Channel Plugin
+ *
+ * Copyright 2019 Mati Shabtay <matishabtay@gmail.com>
+ *
+
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <freerdp/types.h>
+#include <freerdp/constants.h>
+
+#include <freerdp/channels/log.h>
+
+#include <winpr/crt.h>
+#include <winpr/synch.h>
+#include <winpr/thread.h>
+#include <winpr/stream.h>
+
+#include "rail_main.h"
+
+#define TAG CHANNELS_TAG("rail.server")
+
+/**
+ * Sends a single rail PDU on the channel
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rail_send(RailServerContext* context, wStream* s, ULONG length)
+{
+	UINT status = CHANNEL_RC_OK;
+	ULONG written;
+
+	if (!context)
+		return CHANNEL_RC_BAD_INIT_HANDLE;
+
+	if (!WTSVirtualChannelWrite(context->priv->rail_channel,
+	                            (PCHAR) Stream_Buffer(s),
+	                            length, &written))
+	{
+		WLog_ERR(TAG, "WTSVirtualChannelWrite failed!");
+		status = ERROR_INTERNAL_ERROR;
+	}
+
+	return status;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+UINT rail_send_pdu(RailServerContext* context, wStream* s, UINT16 orderType)
+{
+	UINT16 orderLength;
+
+	if (!context || !s)
+		return ERROR_INVALID_PARAMETER;
+
+	orderLength = (UINT16) Stream_GetPosition(s);
+	Stream_SetPosition(s, 0);
+	rail_write_pdu_header(s, orderType, orderLength);
+	Stream_SetPosition(s, orderLength);
+	WLog_DBG(TAG, "Sending %s PDU, length: %"PRIu16"",
+	           RAIL_ORDER_TYPE_STRINGS[((orderType & 0xF0) >> 3) + (orderType & 0x0F)], orderLength);
```

Comment:
what are those cryptic operations ? can it be done in a  cleaner way?

Reply:
They aren't that cryptic IMO, this is how they do it in client side too. I want to merge some of this functionality by exporting to common and reusing in both sides, might clean up when I do that. Just want to focus on making it work for now